# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
df = pd.read_csv('data/step3_output2.csv')
df.head()

Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939           2500      1        0.0   
1  Eaglecrest Ski Area   Alaska         2600           1540      0        0.0   
2     Hilltop Ski Area   Alaska         2090            294      0        0.0   
3     Arizona Snowbowl  Arizona        11500           2300      0        0.0   
4  Sunrise Park Resort  Arizona        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  Snow Making_ac  \
0          0          2     2       0  ...             1610.0           113.0   
1          0          0     0       0  ...              640.0            60.0   
2          0          0     0       1  ...               30.0            30.0   
3          1          0     2       2  ...              777.0           104.0   
4          0          1     2       3  ...              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0            669.0          65.0          85.0   
1              45.0       44.0            350.0          47.0          53.0   
2             150.0       36.0             69.0          30.0          34.0   
3             122.0       81.0            260.0          89.0          89.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              150.0           550.0         0  
1               90.0             0.0         0  
2              152.0            30.0         0  
3              122.0             0.0         1  
4              104.0            80.0         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [3]:
df = pd.concat([df.drop(['state'], axis=1), pd.get_dummies(df['state'])], axis=1)
df.head()
#df.to_csv("dummy.csv",index=False)
#df.describe().T

Name  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort         3939           2500      1        0.0   
1  Eaglecrest Ski Area         2600           1540      0        0.0   
2     Hilltop Ski Area         2090            294      0        0.0   
3     Arizona Snowbowl        11500           2300      0        0.0   
4  Sunrise Park Resort        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0          0          2     2       0       0  ...             0   
1          0          0     0       0       4  ...             0   
2          0          0     0       1       0  ...             0   
3          1          0     2       2       1  ...             0   
4          0          1     2       3       1  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [4]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)
# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [5]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [6]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [7]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = lm.predict(X_test)
#pd.DataFrame({'Actual': y_test.flatten(), 'Predicted': y_pred.flatten()})

In [8]:
print(np.round(y_pred,2))
#pd.DataFrame(X_train).to_csv("test_output.csv")
#pd.DataFrame(X_test).describe().T

[ 57.94  89.49  78.63  75.74  69.21  76.28  44.64  46.77  93.65  84.32
  50.61  79.25  50.7   70.93  62.54  93.38  42.27  73.05  61.89  40.81
 113.6   36.86  53.7   49.4   47.96  63.49  76.28  75.7   63.97  66.6
  34.53  59.1   80.26  57.11  75.53  49.35  22.95  29.22  61.49  37.21
  48.47  63.31  86.63  72.39  72.62  58.47  60.79 165.03  37.77  61.09
  68.41  60.29  53.29  43.78  59.24  42.23  85.5   39.49 119.12 118.98
  56.15  53.66  41.78  80.97  64.62  43.3   36.41  96.76  61.25  24.
  68.14 146.14  57.96  62.39  75.33  71.    76.    58.64  85.3   71.43
  52.47  56.87  43.19]


## Review Model Outcomes — Iterate over additional models as needed

In [9]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

In [10]:
evs = explained_variance_score(y_test,y_pred)
mae = mean_absolute_error(y_test,y_pred)
print(round(evs,4))
print(round(mae,4))

0.9407
4.8857


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [11]:
round(lm.intercept_,4)

64.1026

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [12]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(["Coefficient"],axis=0,ascending = False)

Coefficient
total_chairs       3.130047e+13
fastQuads          1.186608e+13
surface            1.111763e+13
double             9.797264e+12
triple             8.739832e+12
quad               7.083310e+12
New York           6.752307e+12
Michigan           6.372345e+12
Colorado           5.614405e+12
California         5.494218e+12
Pennsylvania       5.242929e+12
New Hampshire      4.834392e+12
Wisconsin          4.834392e+12
Vermont            4.688327e+12
Minnesota          4.536538e+12
Utah               4.378429e+12
Idaho              4.213288e+12
Montana            4.213288e+12
Massachusetts      4.040254e+12
Washington         3.858265e+12
Oregon             3.858265e+12
Maine              3.665986e+12
New Mexico         3.665986e+12
fastSixes          3.517705e+12
Wyoming            3.461704e+12
trams              3.022510e+12
North Carolina     3.007220e+12
Connecticut        2.749436e+12
Ohio               2.749436e+12
West Virginia      2.462951e+12
Virginia           2.462951e+12
Illinois           2.462951e+12
Nevada             2.462951e+12
Alaska             2.136247e+12
Iowa               2.136247e+12
New Jersey         1.746904e+12
Missouri           1.746904e+12
Arizona            1.746904e+12
Indiana            1.746904e+12
South Dakota       1.746904e+12
Tennessee          1.237129e+12
Maryland           1.237129e+12
Rhode Island       1.237129e+12
fastEight          2.971423e+11
AdultWeekday       1.927082e+01
summit_elev        1.708352e+00
Runs               1.278368e+00
vertical_drop      1.110942e+00
Snow Making_ac     9.047742e-01
NightSkiing_ac     7.176418e-01
clusters           6.636573e-01
daysOpenLastYear   4.996071e-01
LongestRun_mi      4.866431e-01
yearsOpen          3.609370e-01
SkiableTerrain_ac  1.120148e-01
projectedDaysOpen  9.617321e-02
averageSnowfall    7.715831e-02
TerrainParks       6.260008e-02

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [13]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

In [14]:
df = pd.read_csv('data/step3_output2.csv')
df = df.drop(['state'], axis=1)
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)
# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)


# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = lm.predict(X_test)

evs = explained_variance_score(y_test,y_pred)
mae = mean_absolute_error(y_test,y_pred)
print("Explained Variance Score is", round(evs,4))
print("Mean Absolute Error is", round(mae,4))
print("Intercept of Linear Model is", round(lm.intercept_,4))
pd.options.display.float_format = '{:,.4f}'.format
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(["Coefficient"],axis=0,ascending = False)

Explained Variance Score is 0.9348
Mean Absolute Error is 5.1261
Intercept of Linear Model is 64.0951


Coefficient
AdultWeekday           19.8417
clusters                5.9697
summit_elev             4.6522
Runs                    1.9296
surface                 1.4353
triple                  1.3775
averageSnowfall         1.0819
quad                    1.0745
vertical_drop           1.0221
fastQuads               0.9789
NightSkiing_ac          0.7066
daysOpenLastYear        0.6067
Snow Making_ac          0.5959
total_chairs            0.5730
LongestRun_mi           0.5233
SkiableTerrain_ac       0.4339
fastSixes               0.2802
yearsOpen               0.2769
fastEight               0.2245
double                  0.1998
projectedDaysOpen       0.1913
TerrainParks            0.1475
trams                   0.1271

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [15]:
df = pd.read_csv('data/step3_output2.csv')
df = df.drop(['state','summit_elev'], axis=1)
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)
# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)


# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = lm.predict(X_test)

evs = explained_variance_score(y_test,y_pred)
mae = mean_absolute_error(y_test,y_pred)
print("Explained Variance Score is", round(evs,4))
print("Mean Absolute Error is", round(mae,4))
print("Intercept of Linear Model is", round(lm.intercept_,4))
pd.options.display.float_format = '{:,.4f}'.format
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(["Coefficient"],axis=0,ascending = False)

Explained Variance Score is 0.9354
Mean Absolute Error is 5.1303
Intercept of Linear Model is 64.0775


Coefficient
AdultWeekday           20.0775
clusters                2.6385
vertical_drop           2.0324
Runs                    1.3986
triple                  1.3870
surface                 1.2490
quad                    1.1839
daysOpenLastYear        1.1567
averageSnowfall         0.8995
fastQuads               0.7052
projectedDaysOpen       0.5851
Snow Making_ac          0.5846
total_chairs            0.5717
LongestRun_mi           0.4038
fastSixes               0.3830
NightSkiing_ac          0.3818
trams                   0.2909
double                  0.2469
yearsOpen               0.2427
SkiableTerrain_ac       0.1416
TerrainParks            0.0925
fastEight               0.0600

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.9407 | 4.8857 |-|
| Model 2. | 0.9348 | 5.1261|'state'|
| Model 3. | 0.9354 | 5.1303 |'state','summit_elev','base_elev'|

Model Selection:

I will choose Model 2 since it has smaller "Mean Absolute Error" than Model 3. Explained Vaiances are very close for model 2 and 3. <br>
Although model 1 has the highest "Explained Variance" and smallest "Mean Absolute Error", the coefficients are extreamly large (this is a very strange result!) So I will eliminated model 1 at first.